File: common-core/src/main/java/io/quartic/common/client/WebsocketClientSessionFactory.java
package io.quartic.common.client;

import org.glassfish.tyrus.client.ClientManager;
import org.glassfish.tyrus.client.ClientProperties;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import javax.websocket.ClientEndpointConfig;
import javax.websocket.CloseReason;
import javax.websocket.DeploymentException;
import javax.websocket.Endpoint;
import javax.websocket.Session;
import java.io.IOException;
import java.net.URI;
import java.net.URISyntaxException;
import java.util.List;
import java.util.Map;

import static com.google.common.collect.Lists.newArrayList;
import static com.google.common.net.HttpHeaders.USER_AGENT;

public class WebsocketClientSessionFactory {
    private static final Logger LOG = LoggerFactory.getLogger(WebsocketClientSessionFactory.class);

    public final Class<?> owner;

    public WebsocketClientSessionFactory(Class<?> owner) {
        this.owner = owner;
    }

    public Session create(Endpoint endpoint, String url) throws URISyntaxException, IOException, DeploymentException {
        return clientManager(url)
                .connectToServer(endpoint, websocketClientConfig(), new URI(url));
    }

    private ClientEndpointConfig websocketClientConfig() {
        return ClientEndpointConfig.Builder.create()
                .configurator(new ClientEndpointConfig.Configurator() {
                    @Override
                    public void beforeRequest(Map<String, List<String>> headers) {
                        headers.put(USER_AGENT, newArrayList(Utils.userAgentFor(owner)));
                    }
                })
                .build();
    }

    private ClientManager clientManager(String url) {
        ClientManager clientManager = ClientManager.createClient();
        ClientManager.ReconnectHandler reconnectHandler = new ClientManager.ReconnectHandler() {

            @Override
            public boolean onDisconnect(CloseReason closeReason) {
                LOG.warn("[{}] Disconnecting: {}", url, closeReason);
                return true;
            }

            @Override
            public boolean onConnectFailure(Exception exception) {
                LOG.warn("[{}] Connection failure: {}\n{}",
                        url,
                        exception.getMessage(),
                        formatStackTrace(exception.getStackTrace()));
                return true;
            }

            @Override
            public long getDelay() {
                return 5;
            }
        };
        clientManager.getProperties().put(ClientProperties.RECONNECT_HANDLER, reconnectHandler);
        return clientManager;
    }

    private String formatStackTrace(StackTraceElement[] stackTrace) {
        final StringBuilder sb = new StringBuilder();
        for (int i = 0; i < Math.min(3, stackTrace.length); i++) {
            sb.append("\tat " + stackTrace[i] + "\n");
        }
        return sb.append("\t...").toString();
    }
}

File: common-core/src/main/kotlin/io/quartic/common/websocket/WebsocketClientSessionFactory.kt
package io.quartic.common.websocket

import io.quartic.common.logging.logger
import org.glassfish.tyrus.client.ClientManager
import org.glassfish.tyrus.client.ClientManager.ReconnectHandler
import org.glassfish.tyrus.client.ClientProperties.RECONNECT_HANDLER
import org.glassfish.tyrus.core.TyrusSession
import java.net.URI
import javax.websocket.CloseReason
import javax.websocket.Endpoint
import javax.websocket.Session

class WebsocketClientSessionFactory @JvmOverloads constructor(
        private val owner: Class<*>,
        private val clientManagerSupplier: () -> ClientManager = { ClientManager.createClient() }
) {
    private val LOG by logger()

    fun create(endpoint: Endpoint, url: String): Session {
        val session = clientManager(url).connectToServer(endpoint, clientEndpointConfig(owner), URI(url)) as TyrusSession
        session.heartbeatInterval = HEARTBEAT_INTERVAL_MILLISECONDS
        return session
    }

    // TODO: can we have a global ClientManager somehow?
    private fun clientManager(url: String): ClientManager {
        val clientManager = clientManagerSupplier()
        clientManager.properties[RECONNECT_HANDLER] = reconnectHandler(url)
        return clientManager
    }

    private fun reconnectHandler(url: String) = object : ReconnectHandler() {
        override fun onDisconnect(closeReason: CloseReason): Boolean {
            LOG.warn("[$url] Disconnecting: $closeReason")
            return true
        }

        override fun onConnectFailure(exception: Exception): Boolean {
            LOG.warn("[$url] Connection failure", exception)
            return true
        }

        override fun getDelay(): Long = 5
    }

    companion object {
        val HEARTBEAT_INTERVAL_MILLISECONDS = 30000L
    }
}
